worksheet.rb

```diff
@@ -23,12 +38,209 @@
 # into this data structure, such as "DR0004"
 # and "3rd Feb 2016" and "RD0022"
 
+ride_share_data = {
+    rides: [
+        {
+            driver_id: 'DR0004',
+            date: '3rd Feb 2016',
+            cost: 5,
+            rider_id: 'RD0022',
+            rating: 5
+        },
+        {
+            driver_id: 'DR0001',
+            date: '3rd Feb 2016',
+            cost: 10,
+            rider_id: 'RD0003',
+            rating: 3
+        },
+        {
+            driver_id: 'DR0002',
+            date: '3rd Feb 2016',
+            cost: 25,
+            rider_id: 'RD0073',
+            rating: 5
+        },
+        {
+            driver_id: 'DR0001',
+            date: '3rd Feb 2016',
+            cost: 30,
+            rider_id: 'RD0015',
+            rating: 4
+        },
+        {
+            driver_id: 'DR0003',
+            date: '4th Feb 2016',
+            cost: 5,
+            rider_id: 'RD0066',
+            rating: 5
+        },
+        {
+            driver_id: 'DR0004',
+            date: '4th Feb 2016',
+            cost: 10,
+            rider_id: 'RD0022',
+            rating: 4
+        },
+        {
+            driver_id: 'DR0002',
+            date: '4th Feb 2016',
+            cost: 15,
+            rider_id: 'RD0013',
+            rating: 1
+        },
+        {
+            driver_id: 'DR0003',
+            date: '5th Feb 2016',
+            cost: 50,
+            rider_id: 'RD0003',
+            rating: 2
+        },
+        {
+            driver_id: 'DR0002',
+            date: '5th Feb 2016',
+            cost: 35,
+            rider_id: 'RD0066',
+            rating: 3
+        },
+        {
+            driver_id: 'DR0004',
+            date: '5th Feb 2016',
+            cost: 20,
+            rider_id: 'RD0073',
+            rating: 5
+        },
+        {
+            driver_id: 'DR0001',
+            date: '5th Feb 2016',
+            cost: 45,
+            rider_id: 'RD0003',
+            rating: 2
+        }
+    ]
+}
+
 ########################################################
 # Step 4: Total Driver's Earnings and Number of Rides
 
 # Use an iteration blocks to print the following answers:
 # - the number of rides each driver has given
+def num_rides(structured_data)
+  driver_ride_count = {}
+  structured_data[:rides].each do |ride|
```

Comment:
Note that `num_rides`, `driver_earnings`, and `average_ratings` all have this loop `structured_data[:rides].each do |ride|`. Consider how you could combine this functionality to reduce repeated code. It a delicate balance between writing many methods so that each method only does one thing and writing fewer methods to avoid repeated code.